src/include/plan_node/abstract_plan_node.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "common/hash_util.h"
+#include "common/json.h"
+#include "plan_node/output_schema.h"
+#include "plan_node/plan_node_defs.h"
+
+// TODO(Gus,Wen): Add equaility operator and hash function support for output_schema
+
+namespace terrier::plan_node {
+//===--------------------------------------------------------------------===//
+// Abstract Plan Node
+//===--------------------------------------------------------------------===//
+
+class AbstractPlanNode {
+ public:
+  explicit AbstractPlanNode(std::shared_ptr<OutputSchema> output_schema) : output_schema_(std::move(output_schema)) {}
+
+  // For Deserialization and DDL statements
+  AbstractPlanNode() = default;
+
+  virtual ~AbstractPlanNode() = default;
+
+  //===--------------------------------------------------------------------===//
+  // Children Helpers
+  //===--------------------------------------------------------------------===//
+
+  void AddChild(std::unique_ptr<AbstractPlanNode> &&child) { children_.emplace_back(std::move(child)); }
+
+  const std::vector<std::unique_ptr<AbstractPlanNode>> &GetChildren() const { return children_; }
+
+  size_t GetChildrenSize() const { return children_.size(); }
+
+  const AbstractPlanNode *GetChild(uint32_t child_index) const {
+    TERRIER_ASSERT(child_index < children_.size(),
+                   "index into children of plan node should be less than number of children");
+    return children_[child_index].get();
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Accessors
+  //===--------------------------------------------------------------------===//
+
+  // Each sub-class will have to implement this function to return their type
+  // This is better than having to store redundant types in all the objects
+  virtual PlanNodeType GetPlanNodeType() const { return PlanNodeType::ABSTRACTPLAN; }
```

Comment:
You probably want this to be a pure virtual method since I don't think anything will have most specific type of abstract plan. There are pure virtual methods in this class already.